core/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java

```diff
@@ -151,9 +147,17 @@ public ExecutorScalingQueue() {
 
         @Override
         public boolean offer(E e) {
+            // first try to transfer to a waiting worker thread
             if (!tryTransfer(e)) {
+                // check if there might be spare capacity in the thread
+                // pool executor
```

Comment:
> AFAICS the expression `executor.getMaximumPoolSize() - executor.getCorePoolSize()` is fixed once we have constructed the threadpool

@ywelsch This is correct and was already taken into consideration in my assessment of the situation.

From [`java.util.concurrent.ThreadPoolExecutor#execute`](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/concurrent/ThreadPoolExecutor.java#l1332):

```java
        int c = ctl.get();
        if (workerCountOf(c) < corePoolSize) {
            if (addWorker(command, true))
                return;
            c = ctl.get();
        }
        if (isRunning(c) && workQueue.offer(command)) {
            int recheck = ctl.get();
            if (! isRunning(recheck) && remove(command))
                reject(command);
            else if (workerCountOf(recheck) == 0)
                addWorker(null, false);
        }
        else if (!addWorker(command, false))
            reject(command);
```

The first thing that the `ThreadPoolExecutor` does is checks if there are fewer than core pool size threads. If there are, it tries to add a new worker to handle the submitted task. If this fails (it's racy!), `ThreadPoolExecutor` then tries to place the task on the queue. This is where we turn to the `ExecutorScalingQueue`:

```java
            public boolean offer(E e) {
            if (!tryTransfer(e)) {
                int left = executor.getMaximumPoolSize() - executor.getCorePoolSize();
                if (left > 0) {
                    return false;
                } else {
                    return super.offer(e);
                }
            } else {
                return true;
            }
        }
```

We first try to transfer the task to an available worker thread. If this fails, we come to the code in question. We check if there *might* be spare capacity in the thread pool, which can only be the case if the thread pool was configured with max pool size greater than core pool size. If there *might* be spare capacity, we return `false`. This causes the `ThreadPoolExecutor` to create a new worker thread, even if core pool size threads are already running. Again, the `ThreadPoolExecutor` prefers queueing to creating new workers when there are core pool size threads running, and will only resort to creating a new thread above core pool size threads if the queue is saturated which we faked by returning `false`. If this fails (again, it's racy), `ThreadPoolExecutor` will reject the task. This is where the `ForceQueuePolicy` rejection handler comes into play to give us a chance to just drop this on the queue.

I continue to disagree with your and @bleskes assessment of the situation. It is not an optimization, it is a trick for the common case to make the `ThreadPoolExecutor` spin up threads all the way up to max pool size instead of queuing.